Functional_test

mre opened the pull request:
The functional test is based on
https://github.com/Klaus2m5/6502_65C02_functional_tests

It does NOT pass yet. :(
However, I think it's a good idea to have it in the repo
nonetheless, to make some incremental improvements.

This PR is based on the `asm` branch, which should be
merged first.

mre commented on src/cpu.rs:
```diff
@@ -988,11 +988,7 @@ impl<M: Bus> CPU<M> {
-            "CPU Dump:\n\nAccumulator: {}",
-            self.registers.accumulator
-        )
+        write!(f, "CPU {{ registers: {:?}", self.registers)
```
I think printing all registers makes debugging easier.
Not sure if we also want to dump memory. Probably not. It's "static" anyway.

mre replied:
Example output:

```
CPU { registers: Registers { accumulator: 0, index_x: 255, index_y: 0, stack_pointer: StackPointer(255), program_counter: 1529, status: Status(PS_OVERFLOW | PS_UNUSED | PS_BRK | PS_DECIMAL_MODE | PS_DISABLE_INTERRUPTS | PS_ZERO | PS_CARRY) }
CPU { registers: Registers { accumulator: 0, index_x: 255, index_y: 0, stack_pointer: StackPointer(254), program_counter: 1530, status: Status(PS_OVERFLOW | PS_UNUSED | PS_BRK | PS_DECIMAL_MODE | PS_DISABLE_INTERRUPTS | PS_ZERO | PS_CARRY) }
CPU { registers: Registers { accumulator: 0, index_x: 255, index_y: 0, stack_pointer: StackPointer(255), program_counter: 1531, status: Status(0x0) }
CPU { registers: Registers { accumulator: 0, index_x: 255, index_y: 0, stack_pointer: StackPointer(255), program_counter: 1533, status: Status(0x0) }
CPU { registers: Registers { accumulator: 0, index_x: 255, index_y: 0, stack_pointer: StackPointer(255), program_counter: 1535, status: Status(0x0) }
```

omarandlorraine commented:
>I don't know what's the issue yet.

It could it decimal arithmetic on invalid input BCD. As far as I know, that's the only issue solid65 finds with mre/6502.

As far as I can see, the functional test has some kind of conditional compilation to enable/disable decimal mode tests. Does the test pass if you disable this?

omarandlorraine commented on examples/asm/functional_test/README.md:
I think it could do with a description of precisely what the test is.

Which sourcefile is this binary built from? Which conditional compilation flags have been enabled?

mre replied:
Good idea. I've extended the description a bit.

mre commented on examples/functional.rs:
```diff
@@ -0,0 +1,34 @@
+        cpu.single_step();
+        println!("{cpu:?}");
+
+        if cpu.registers.program_counter == old_pc {
```
To be honest, I blindly took it from here:
https://github.com/marioballano/emudore/blob/5d98f4447409b1d778b9305b0df9aea064d536c5/src/c64.cpp#L134

They are using the same test image. It might be completely wrong, though. The actual repository for the test rom is a bit sparse on details on how to run the test. At least I couldn't find anyting in their `README.md` and the top of the rom file:

https://github.com/Klaus2m5/6502_65C02_functional_tests
https://github.com/Klaus2m5/6502_65C02_functional_tests/blob/master/6502_functional_test.a65

I could not compile the ROM file myself, because I use ca65 and the rom was built with as65. There seems to be a fork of the original repo, but I didn't dig deeper.

mre replied:
Update: I compiled the functional test from the sources now. I used https://github.com/amb5l/6502_65C02_functional_tests, which is a port of the original code to ca65, which is the assembler I have on my mac. Now we get even less far in the test. 🤣 

Here's all I get:

```asm
INC $4868
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
BNE $FFFE
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
Infinite loop detected!
```

Probably I messed up the compilation or the memory segments.

omarandlorraine replied:
> Update: I compiled the functional test from the sources now. I used https://github.com/amb5l/6502_65C02_functional_tests, which is a port of the original code to ca65, which is the assembler I have on my mac. Now we get even less far in the test. 🤣

Does ca65 have an option to export the labels it knows? Then you could find which infinite loop starts at address 1029 (the reported program counter).

mre replied:
Yes, ca65 has a `-g` flag, which adds debug info to the object file. I added that and also added the flags to create a symbols file (`labels.dbg`). I've updated the Rust code to use the labels &mdash; or at least I hope that's how it works.

I'm not getting any label output, just

```
INC $4868
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
BNE $FFFE
CPU { registers: Registers { accumulator: 0, index_x: 0, index_y: 0, stack_pointer: StackPointer(0), program_counter: 1029, status: Status(PS_NEGATIVE | PS_UNUSED | PS_DISABLE_INTERRUPTS) }
Infinite loop detected!
```

like before. 
I think my setup is wrong and it doesn't even execute anything at the moment.

mre replied:
I'm getting label output now after my latest changes and by adjusting the memory segments in the linker config, I get the following output when I run `cargo run --release --example functional`:

```
CLD
TXS
STA $0200 (.test_case)
LDX #$05
BNE $FFF4
DEX
DEX
BEQ $FFDE
BNE $0008
DEY
DEY
BEQ $0017
CMP #$00
LDA #$01
LDA $0200,X (.test_case)
BNE $FFFE
STA $0200 (.test_case)
BNE $0003
CMP #$00
BCC $FFFE
CMP #$01
BCS $FFFE
TAX
BNE $FFFE
BMI $FFFE
BEQ $FFFE
BPL $FFFE
CPY #$00
BCC $FFFE
CPY #$01
BCS $FFFE
LDA $0200,X (.test_case)
BNE $FFFE
STA $0200 (.test_case)
TXS
PHA
PHA
BNE $FFFE
TXA
BNE $FFFE
CMP #$AA
PLA
BNE $FFFE
BNE $FFFE
CPX #$FF
LDA $0200,X (.test_case)
BNE $FFFE
STA $0200 (.test_case)
PHA
BPL $001A (.zpOR)
BCC $001C
BMI $0003
BCS $0003
PHP
CPX #$FE
PLA
BNE $FFFE
CPX #$FF
LDA #$04
PLP
BVS $001B
BEQ $001D
BVC $0003
BNE $000F (.adrl)
PLA
BNE $FFFE
Infinite loop detected!
CPU { registers: Registers { accumulator: 4, index_x: 255, index_y: 0, stack_pointer: StackPointer(255), program_counter: 1336, status: Status(PS_NEGATIVE | PS_DISABLE_INTERRUPTS) }
```

omarandlorraine commented on examples/asm/euclid/euclid.a65:
```diff
@@ -0,0 +1,33 @@
+; .algo
+LDA $00    ; Load from F to A
+; .algo_
+sec        ; Set carry flag
```
`sec` is randomly in lower case

mre replied:
lol, yeah. good catch.

omarandlorraine reviewed (approved):
LGTM

omarandlorraine commented on src/instruction.rs:
```diff
@@ -152,7 +154,24 @@ impl AddressingMode {
+            OpInput::UseImplied => write!(f, "{:?}", self.0),
+            OpInput::UseImmediate(v) => write!(f, "{:?} #${:02X}", self.0, v),
+            OpInput::UseRelative(v) => write!(f, "{:?} ${:04X}", self.0, v),
+            OpInput::UseAddress(v) => write!(f, "{:?} ${:04X}", self.0, v),
```
How does this code handle indexed and indirect addressing modes? I would expect format strings for things like `adc ($1234),x` or `adc ($12),y` for example.

mre replied:
It doesn't at the moment. This just impls Display for `DecodedInstr`, which is defined as
```rust
pub struct DecodedInstr(pub Instruction, pub OpInput);
```

where `OpInput` is 

```rust
pub enum OpInput {
    UseImplied,
    UseImmediate(u8),
    UseRelative(u16),
    UseAddress(u16),
}
```

We don't have the addressing mode in there. 😢 Instead, there is 

```rust
pub static OPCODES: [Option<(Instruction, AddressingMode)>; 256] = [
    /*0x00*/
    Some((Instruction::BRK, AddressingMode::Implied)),
    /* ... */
```

which does the mapping to the `AddressingMode`. I've added that to the print output. Can you have another look to see if it's better now and covers all of our cases?

omarandlorraine reviewed (commented):
Do you know why I'm seeing this message?

`function `decimal_subtract_test` is never used`

mre commented:
> Do you know why I'm seeing this message?
>    function decimal_subtract_test is never used

No idea. I don't get that message. Is that when you compile the project?

omarandlorraine commented:
> > Do you know why I'm seeing this message?
> > function decimal_subtract_test is never used
> 
> No idea. I don't get that message. Is that when you compile the project?

No, it's in the code review area on Github. But I can't find it again now 🤷

mre commented:
> EDIT: Oh here it is. On line 1044 of cpu.rs.

Okay, no idea. Looks like a false-positive to me. I don't see that message if I scroll to that line.

mre commented:
I fixed quite a few critical bugs in the 6502 emulator's stack operations that were preventing the Klaus2m5 functional test from running correctly. The journey started when I rebased the functional_test branch onto master, which brought in the recent SBC improvements from PR #117. This immediately showed progress. The test went from failing at PC ~1200 to reaching much further into the test suite.

However, the test was still hitting infinite loops and illegal opcodes. After careful analysis of the execution trace, I discovered that the RTI (Return from Interrupt) and RTS (Return from Subroutine) instructions had a fundamental bug: they were calling pull_from_stack() to discard a byte, then using fetch_from_stack() to read the actual data. This pattern was working around an even deeper bug in the pull_from_stack() method itself, which was reading from the stack pointer and then incrementing it. The 6502's stack pointer actually points to the next available slot, so pulls should increment first, then read. Once I fixed pull_from_stack() to use the correct order, I had to update RTI and RTS to simply call pull_from_stack() directly without the workaround.

The same bug pattern appeared in all the pull instructions—PLA, PLP, PLX, and PLY were all doing the pull-then-fetch dance. I fixed each of them to use a single pull_from_stack() call. I also simplified the Display implementation for DecodedInstr, which was trying to reference a non-existent OPCODES table that had been removed from the codebase long ago. After these fixes, the test runs for about 30 million instructions before hitting a trap, which is a massive improvement from the original ~600 instructions. 🎉 

Finally, I restructured the test suite itself. I converted the example into a proper integration test in `tests/functional_test.rs`, replacing all the magic numbers with named constants like SUCCESS_ADDR (0x3468) and PROGRAM_START_ADDR (0x0400). I moved the test assets from examples/asm/functional_test/ to tests/assets/ where they belong, and rewrote the test to use assertions instead of panics for better test failure messages. The old example is gone, and we now have a clean integration test that will fail fast with clear error messages when something goes wrong.

And there is still an issue somewhere. Will investigate more. But overall, I believe we should merge it soon as it's a significant improvement already IMHO.
